registrations/views.py

```diff
@@ -47,13 +53,46 @@ def get_context_data(self, **kwargs):
         context["clinic_name"] = self.request.session["clinic_name"]
         return context
 
+    def get_channel(self, msisdn):
+        """
+        Returns "WhatsApp" or "SMS" depending on whether the MSISDN is
+        whatsappable.
+
+        Args:
+            msisdn (str): The MSISDN to query
+        """
+        try:
+            wabclient.get_address(msisdn)
+            return "WhatsApp"
+        except AddressException:
+            return "SMS"
+
     def post(self, request, *args, **kwargs):
-        if "yes" in request.POST:
-            # TODO: WhatsApp check
-            # TODO: Create registration
-            request.session["channel"] = "WhatsApp"
-            return redirect(reverse_lazy("registrations:success"))
-        return redirect(reverse_lazy("registrations:registration-details"))
+        if "yes" not in request.POST:
+            return redirect(reverse_lazy("registrations:registration-details"))
+
+        # TODO: Create registration
+
+        try:
+            request.session["channel"] = self.get_channel(
+                request.session["registration_details"]["msisdn"]
+            )
+        except RequestException:
+            # Ensure that we know of repeating errors
+            try:
+                request.session["whatsapp_api_errors"] += 1
+            except KeyError:
+                request.session["whatsapp_api_errors"] = 1
+            if request.session.get("whatsapp_api_errors", 0) >= 3:
```

Comment:
Why do we want to track this on a session? wouldn't we want to track it across multiple users so that we can more reliably be alerted if something is wrong?

Reply:
So we know that this API goes down for a few seconds every time there's a deploy. There's nothing that we can do about that, since we can't do retries since it's in the view, so we have to rely on the user clicking a button for our retries. Since it's a known case that we can't do anything about, we don't want to alert every time it fails.

The problem with tracking this globally is that if it goes down when there are lots of users trying to register at the same time, then we'll easily hit whatever limit we set, even though it's a known case that we don't want in our alerts. We don't really want to get an alert if there are a hundred users that get a single failure, but only get an alert after a single user retries 100 times.

What we do want to know is, if it's down for longer than short periods. A user retrying twice in a row is (thumbsucked) probably around the point when they'll give up, so that's something that we want to be alerted on.

Thinking further about this though, this is probably something that we'll want to track. I'll add a prometheus metric here to track the failures over time.

Reply:
ok, makes sense :+1: